### ISSUE TITLE
Should the participant identifier be an integer or 32-byte array?

### ISSUE THREAD

**Issue by siv2r** [siv2r/bip-frost-signing#9]
Currently, all the high-level functions/objects—such as `sign`, `derive_group_pubkey`, `nonce_agg`, and `session context`—use 32-byte arrays to represent the participant identifier. These functions internally convert the byte arrays to integers whenever they need to compute the Lagrange coefficient or blame a disruptive signer.

Theoretically, a 32-byte array is required for the participant identifier since it represents a scalar value (< curve order). However, in practice, the `t`-of-`n` values would not be that large, so the identifier could be represented as a 4 or 8-byte integer instead.
<br>

> identifier could be represented as a 4 or 8-byte integer instead.  

This is only possible when $id < \text{MAXPARTICIPANTS}$. If we adopt a design where IDs can be random (see this https://github.com/siv2r/bip-frost-signing/issues/5#issuecomment-2175475776), we must stick to using a 32-byte array.

**Comment by real-or-random** [comment]
We decided on integer indices as ids for ChillDKG now (see https://github.com/BlockstreamResearch/bip-frost-dkg/pull/64). 

Unless you want to make the signing BIP more generic and support arbitrary ids (e.g., if a different key generation method outputs them), taking integers seems much more natural to me than 32-byte arrays.

**Comment by siv2r** [comment]
I am inclined to use integers ids, too. Since the signing BIP does not use arbitrary id values (it uses 1..n), the ids are directly tied to the length of the `pubshare` list. This list is unlikely to require large values (e.g., 32 bytes). Therefore, using 4-byte integers makes sense. I'll make this change soon.

It is worth noting that ChillDKG signers must increment their `hostpubkey` list indices by one before using them as ids in the signing BIP.

**Comment by real-or-random** [comment]
> It is worth noting that ChillDKG signers must increment their `hostpubkey` list indices by one before using them as ids in the signing BIP.

Yeah, I've just noticed the same a few minutes ago... I think this is pretty inconvenient for the user, and we should try to get rid of that requirement. But not sure what the best way is. cc @jonasnick

**Comment by siv2r** [comment]
Incrementing ids in FROST Signing BIP might look unusual and could potentially lead to incompatibility with other key generation protocols.

The FROST RFC ([9791](https://datatracker.ietf.org/doc/rfc9591/)) `sign` function specifies:
>   Inputs:
>   - identifier, identifier i of the participant, a NonZeroScalar.


and it uses the id directly to compute the lagrange coefficient

>    def sign(identifier, sk_i, group_public_key,
>             nonce_i, msg, commitment_list):
>      .....
>      .....
>      lambda_i = derive_interpolating_value(participant_list, identifier)


Should this increment be addressed when choosing the signer set? That will happen outside of these two BIPs, right?

**Comment by jonasnick** [comment]
@siv2r If you're planning to use 4-byte indices as ids then denoting the first id as 0 would be the natural choice and also what users would expect. Ids starting at 1 would be yet another thing users would have to consider. That `f(0)` is not a secret share is an implementation detail and should not leak to the API (if ids are indices).

I don't see how this would introduce new incompatibilities to the RFC: ids `1` to `n` and accessing the polynomial at `f(id)` is equivalent to ids `0` to `n-1` and `f(id + 1)`. Aren't we incompatible in other ways with the RFC already (e.g. by using FROST 3)?

Alternatively, if you want ids to start at 1, BIP FROST-signing could explicitly refer to BIP DKG and mention that it's compatible (I believe we want this anyway). Then define that the "id" is ChillDKG index + 1.

**Comment by siv2r** [comment]
> ids `1` to `n` and accessing the polynomial at `f(id)` is equivalent to ids `0` to `n-1` and `f(id + 1)`

Yes, I agree. I had been thinking of ids as `1..n` for so long that incrementing them within the functions felt a bit strange to me at first glance.

> Aren't we incompatible in other ways with the RFC already (e.g. by using FROST 3)?

When I referred to incompatibilities, I meant key generation protocols that could use ids `1..n` (though none currently exist). You're right—sticking to the same RFC design doesn’t make much sense since we’ve already deviated with FROST3.

I'll update the BIP to use 4-byte integers indices (`0..n-1`) for ids.

**Comment by Christewart** [comment]
> If you're planning to use 4-byte indices as ids then denoting the first id as 0 would be the natural choice and also what users would expect. Ids starting at 1 would be yet another thing users would have to consider. That `f(0)` is not a secret share is an implementation detail and should not leak to the API (if ids are indices).

Coming [from this](https://github.com/bitcoin/bips/pull/2070#discussion_r2841067930)

>denoting the first id as 0 would be the natural choice and also what users would expect

I think this actually introduces more complexity and ambiguity. From implementing [BIP445](https://github.com/siv2r/bips/blob/289286c95bd719cdd0fb3c2d3ad19a64193d614c/bip-0445.md) along with [RFC9571's trusted dealer protocol](https://www.rfc-editor.org/rfc/rfc9591.html#name-trusted-dealer-key-generati) we now need to keep different mental models of what are "ids" (rather, indexes?) and if they need to be shifted. I thought copilot wrote a [thoughtful critique when code reviewing my implementation](https://github.com/bitcoin-s/bitcoin-s/pull/6190#discussion_r2836581731). Unfortunately copilot isn't smart enough to realize that VSS commitments don't work with `id=0`. If copilot doesn't understand this, it seems like there will be substantial risk that other developers don't understand this.

I also find it harder to implement proper invariants through the codebase ("which places are `id=0` allowed, and which are not?") which could lead to subtle bugs in implementations when handling key generation protocols and then signing/verification.

Sorry - this comment probably isn't constructive. Just airing my grievances before moving on. Admittedly I haven't read ChillDKG yet - so things may make more sense after that. However its unfortunate to have this incompatibility with simple trusted dealer setups.

**Comment by siv2r** [comment]
> From implementing [BIP445](https://github.com/siv2r/bips/blob/289286c95bd719cdd0fb3c2d3ad19a64193d614c/bip-0445.md) along with [RFC9571's trusted dealer protocol](https://www.rfc-editor.org/rfc/rfc9591.html#name-trusted-dealer-key-generati) we now need to keep different mental models of what are "ids" (rather, indexes?)

There's a [trusted dealer BIP in development](https://github.com/BEULAHEVANJALIN/bip-frost-trusted-dealer) which will follow the "ids as indices" convention. So you won't need to have different conventions within your codebase :)

**Comment by real-or-random** [comment]
Identifiers are integers 0..n-1. This is the only thing that appears in the API and the thing we should primarily care about. Essentially every contemporary programming language uses 0 as the first index for sequences (arrays, lists, vectors, or whatever they're called). 

The fact that the participant i has its secret share at f(i+1) is something that only implementations of FROST need to care about. But this is the cryptographic code itself. I expect everyone implementing cryptography to read the spec and test their code against the test vectors. 

What the signing spec could do is to be more flexible and expect arbitrary scalars as identifiers, possibly encoded as bytestrings. I, personally, don't see much value in this. The caller could always sort them to obtain an order. (Perhaps the signing BIP should say this.) 

> I also find it harder to implement proper invariants through the codebase ("which places are `id=0` allowed, and which are not?")

I think you just need a function that maps ids to scalars. These are different data types, so you need some way to convert integers to scalars anyway. In the BIPs, this function is just `idToScalar(x: unsigned int) = Scalar(x+1)`.
